MAINT: simplifications related to NumPy bounds

tylerjereddy opened the pull request:
* A few simplifications are empowered by the bump in lower NumPy version bound to `1.25.2`. I suspect there are others, but these are the ones I found from some initial `git grep` work.

[skip circle]

tylerjereddy committed:
MAINT: simplifications related to NumPy bounds

* A few simplifications are empowered by the bump in lower
NumPy version bound to `1.25.2`. I suspect there are others,
but these are the ones I found from some initial `git grep`
work.

[skip circle]

lucascolley commented on scipy/_lib/_util.py:
```diff
@@ -22,16 +22,10 @@
+from numpy.exceptions import (
+    AxisError, ComplexWarning, VisibleDeprecationWarning,
+    DTypePromotionError
+)
```
If they are imported just to be re-exported, the typical solution is to add them to `__all__`.

tylerjereddy replied:
Ok, I pushed in some shims to use `__all__` for the linting; I also had to purge some of the old static typing-related shims that `mypy` started complaining about--perhaps those are no longer needed with the simpler import structure.

lucascolley replied:
not that it blocks this PR, but a complete `__all__` for the module would be better than a partial one

rgommers replied:
The better fix is to change to direct imports from numpy. It was only going via this file to avoid the change to `numpy.exceptions`. I can push a fix.

tylerjereddy committed:
MAINT: PR 22489 revisions

* Based on reviewer feedback, use `__all__` to deal with
linter complaints in `_util.py`. Also needed to purge out
some typing shims related to `AxisError`, `ComplexWarning`,
and `VisibleDeprecationWarning` that are now causing
`mypy` complaints.

[skip circle]

rgommers committed:
MAINT: import warnings from numpy.exceptions directly, rather than `_lib`

larsoner reviewed (approved):
Love a good code cleanup PR!
